### PR TITLE
Manage actions cancelling

### PR DESCRIPTION
**Pull request by FabianDevel**
Depends on #488

### CODE REVIEW

**Review by KevinBoulongne — APPROVED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt [hunk 1]
```diff
@@ -291,21 +291,29 @@ class MainViewModel(application: Application) : AndroidViewModel(application) {
...
                 refreshFolders(mailbox, messagesFoldersIds, archiveId)
             } else {
-                return messagesFoldersIds
+                impactedFoldersIds = messagesFoldersIds
             }
         }
 
-        return null
+        val undoDestinationId = message?.folderId ?: currentFolderId.value
+        val undoIds = messages.getFoldersIds(exception = undoDestinationId) + archiveId
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt:305]: Rename to `undoFoldersIds`

File: app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt [hunk 2]
```diff
@@ -342,16 +351,20 @@ class MainViewModel(application: Application) : AndroidViewModel(application) {
             apiResponse.isSuccess()
         }
 
-        showDeleteSnackbar(isSuccess, shouldPermanentlyDelete, message, undoResource)
-
         if (isSuccess) refreshFolders(mailbox, messages.getFoldersIds(exception = trashId), trashId)
+
+        val undoDestinationId = message?.folderId ?: currentFolderId.value
+        val undoIds = (messages.getFoldersIds(exception = undoDestinationId) + trashId).filterNotNull()
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt:357]: Rename to `undoFoldersIds`

File: app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt [hunk 3]
```diff
@@ -465,10 +486,27 @@ class MainViewModel(application: Application) : AndroidViewModel(application) {
...
+            val snackbarTitle = if (ApiRepository.undoAction(resource).data == true) {
+                refreshFolders(mailbox, foldersIds, destinationFolderId)
+                R.string.snackbarMoveCancelled
+            } else {
+                RCore.string.anErrorHasOccurred
+            }
+
+            snackbarFeedback.postValue(context.getString(snackbarTitle) to null)
+        }
+    }
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt:508]:
```suggestion
    fun undoAction(undoData: UndoData) = viewModelScope.launch(Dispatchers.IO) {
        val mailbox = currentMailbox.value ?: return@launch
        val (resource, foldersIds, destinationFolderId) = undoData

        val snackbarTitle = if (ApiRepository.undoAction(resource).data == true) {
            refreshFolders(mailbox, foldersIds, destinationFolderId)
            R.string.snackbarMoveCancelled
        } else {
            RCore.string.anErrorHasOccurred
        }

        snackbarFeedback.postValue(context.getString(snackbarTitle) to null)
    }
```

**Review by sirambd — CHANGES_REQUESTED** [review]

**Comment by sirambd** [comment]
Rename all `undoIds` with `undoFoldersIds`

**Review by sirambd — APPROVED** [review]

**Review by JorisBodin — APPROVED** [review]

**Review by LunarX — APPROVED** [review]

**Pull request merged**